### ISSUE TITLE
FilterAllow and FilterDeny or a rule negation mechanism would be nice.

### ISSUE THREAD

**Issue by tb3088** [tinyproxy/tinyproxy#48]
Or perhaps instead of introducing new keywords and files to parse, perhaps by appending a '!' (separated by a space) to a rule so that it negates the default action.

(aside: Extrapolating further, this concept could be applied to making rules time/day dependent)

Say you're running in FilterDefaultDeny=Yes and you've white-listed 'microsoft.com'. But you want to block 'developer.microsoft.com' just because. Now sure, one can come up with a regex to allow all the zillion other microsoft sub-domains but deny just this one or a handful. However it would be a darn sight more convenient to have a FilterDeny file that gets processed first and short-circuits further evaluation on a hit.

Of course any unintended overlap between FilterDeny and Filter[Allow] can be considered operator error.

In a similar vein, if FilterDefaultDeny=No, blocking 'developer.microsoft.com' is trivial. But what if I really just want to allow 'updates.microsoft.com' out of all sub-domains? FilterAllow would get processed first and likewise short-circuit.

Why, you ask? The world isn't so simple that a binary approach is reasonable. And folks writing the rules are unlikely to be REGEX experts.

**Comment by captn3m0** [comment]
I like the `!` prefix idea. The code already checks the first character for `#` (comments).

**Comment by rofl0r** [comment]
i agree that the current filtering directives are far from perfect, they seem to have been added on an ad-hoc basis without a clear design in mind.
since we have to care about backwards compatibility, we could design and add a new filtering system with a new directive, e.g. FilterNew where rules are prefixed with some directives.
e.g.
```
# mode prefixes: P - plain (fnmatch-style), R - regex, E - extended regex
# permission prefixes: B - blacklist/block, W - whitelist/allow
# block all domains starting with adserver.
PB adserver.*
RW developer.microsoft.com
RB .*\.microsoft\.com$
# etc
```
